Emit variable is set to `any`, breaks build but works fine in dev mode

### Vue version

3.3.4

### Link to minimal reproduction

https://github.com/rodrigocfd/vue-bug-emits

### Steps to reproduce

Run `npm run dev` and everything works fine.

Run `npm run build` and it won't compile.

### What is expected?

A successful build with `npm run build`.

### What is actually happening?

With `npm run build`, the following error shows up:

```
src/App.vue:10:22 - error TS7006: Parameter 'n' implicitly has an 'any' type.

10  <Child @changeName="n => name = n" />
                        ~


Found 1 error in src/App.vue:10
```

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
    Memory: 25.86 GB / 31.95 GB
  Binaries:
    Node: 18.14.1 - C:\Program Files\Nodejs\node.EXE
    npm: 9.6.7 - C:\Program Files\Nodejs\npm.CMD
  Browsers:
    Chrome: 114.0.5735.91
    Edge: Spartan (44.19041.1266.0)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    vue: ^3.3.4 => 3.3.4
```


### Any additional comments?

I tested the original dependencies installed with `npm create vite@latest`, and then updating all dependencies. The behavior is the same.

Curiously, despite running flawlessly in dev mode, when hovering the variable in VSCode, Volar reports the emit parameter as `any`, which is exactly the build error:

![volar-any](https://github.com/vuejs/core/assets/1283868/92a23918-253b-4674-ab42-9d4d13f6c1c7)

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -24,6 +24,8 @@ type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalCo
 // @ts-ignore
 type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
+// Hack: `Type & Type` is a hack to make unresolved imports' type work.
+// Ref: #3274
 type __VLS_GlobalComponents =
 	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>
 	& __VLS_PickNotAny<__VLS_VueRuntimeCoreGlobalComponents & __VLS_VueRuntimeCoreGlobalComponents, {}>

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -17,6 +17,7 @@ type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};
 
+// TODO: revert this once microsoft/TypeScript#54630 is fixed
 // @ts-ignore
 type __VLS_VueGlobalComponents = import('vue').GlobalComponents;
 // @ts-ignore

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -18,14 +18,10 @@ type __VLS_Prettify<T> = {
 } & {};
 
 type __VLS_GlobalComponents =
-	// @ts-ignore
 	__VLS_PickNotAny<import('vue').GlobalComponents & {}, {}>
-	// @ts-ignore
 	& __VLS_PickNotAny<import('@vue/runtime-core').GlobalComponents & {}, {}>
-	// @ts-ignore
 	& __VLS_PickNotAny<import('@vue/runtime-dom').GlobalComponents & {}, {}>
 	& Pick<typeof import('${vueCompilerOptions.lib}'),
-		// @ts-ignore
 		'Transition'
 		| 'TransitionGroup'
 		| 'KeepAlive'

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -25,6 +25,18 @@ type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalCo
 type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
 // Hack: \`Type & Type\` is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
+// Since typescript 5.1.3, the behavior of unresolved imports has changed.
+// For example, the following code:
+// ```ts
+// type A = import('foo').NotExist extends any ? true : false;
+// type B = A extends any ? true : false;
+// ```
+// In typescript 5.0.4, A and B are both `true`.
+// However, in 5.1.3, A becomes `any` and B becomes `any` too!
+// Actually, any type which includes unresolved imports will be turned into `any`.
+// Even only a reference, like B.
+// So we need to turn it into a `real any` and avoid pollution.
+// This hack does it by intersect itself with itself 
 // Ref: #3274
 type __VLS_GlobalComponents =
 	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -27,15 +27,15 @@ type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComp
 // Hack: \`Type & Type\` is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
 // Since typescript 5.1.3, the behavior of unresolved imports has changed.
 // For example, the following code:
-// ```ts
+// \`\`\`ts
 // type A = import('foo').NotExist extends any ? true : false;
 // type B = A extends any ? true : false;
-// ```
-// In typescript 5.0.4, A and B are both `true`.
-// However, in 5.1.3, A becomes `any` and B becomes `any` too!
-// Actually, any type which includes unresolved imports will be turned into `any`.
+// \`\`\`
+// In typescript 5.0.4, A and B are both \`true\`.
+// However, in 5.1.3, A becomes \`any\` and B becomes \`any\` too!
+// Actually, any type which includes unresolved imports will be turned into \`any\`.
 // Even only a reference, like B.
-// So we need to turn it into a `real any` and avoid pollution.
+// So we need to turn it into a \`real any\` and avoid pollution.
 // This hack does it by intersect itself with itself 
 // Ref: #3274
 type __VLS_GlobalComponents =

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -17,12 +17,21 @@ type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};
 
+declare module 'vue' {
+	export interface GlobalComponents {}
+}
+
+declare module '@vue/runtime-core' {
+	export interface GlobalComponents {}
+}
+
+declare module '@vue/runtime-dom' {
+	export interface GlobalComponents {}
+}
+
 type __VLS_GlobalComponents =
-	// @ts-ignore
-	__VLS_PickNotAny<import('vue').GlobalComponents, {}>
-	// @ts-ignore
+	& __VLS_PickNotAny<import('vue').GlobalComponents, {}>
 	& __VLS_PickNotAny<import('@vue/runtime-core').GlobalComponents, {}>
-	// @ts-ignore
 	& __VLS_PickNotAny<import('@vue/runtime-dom').GlobalComponents, {}>
 	& Pick<typeof import('${vueCompilerOptions.lib}'),
 		// @ts-ignore

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -17,22 +17,17 @@ type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};
 
-declare module 'vue' {
-	export interface GlobalComponents {}
-}
-
-declare module '@vue/runtime-core' {
-	export interface GlobalComponents {}
-}
-
-declare module '@vue/runtime-dom' {
-	export interface GlobalComponents {}
-}
+// @ts-ignore
+type VueGlobalComponents = import('vue').GlobalComponents;
+// @ts-ignore
+type VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalComponents;
+// @ts-ignore
+type VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
 type __VLS_GlobalComponents =
-	& __VLS_PickNotAny<import('vue').GlobalComponents, {}>
-	& __VLS_PickNotAny<import('@vue/runtime-core').GlobalComponents, {}>
-	& __VLS_PickNotAny<import('@vue/runtime-dom').GlobalComponents, {}>
+	& __VLS_PickNotAny<VueGlobalComponents & VueGlobalComponents, {}>
+	& __VLS_PickNotAny<VueRuntimeCoreGlobalComponents & VueRuntimeCoreGlobalComponents, {}>
+	& __VLS_PickNotAny<VueRuntimeDomGlobalComponents & VueRuntimeDomGlobalComponents, {}>
 	& Pick<typeof import('${vueCompilerOptions.lib}'),
 		// @ts-ignore
 		'Transition'

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -7,8 +7,8 @@ export function getTypesCode(vueCompilerOptions: VueCompilerOptions) {
 	return `
 // @ts-nocheck
 
-type __VLS_IntrinsicElements = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.IntrinsicElements, __VLS_PickNotAny<JSX.IntrinsicElements, Record<string, any>>>;
-type __VLS_Element = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.Element, JSX.Element>;
+type __VLS_IntrinsicElements = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.IntrinsicElements & {}, __VLS_PickNotAny<JSX.IntrinsicElements, Record<string, any>>>;
+type __VLS_Element = __VLS_PickNotAny<import('vue/jsx-runtime').JSX.Element & {}, JSX.Element>;
 
 type __VLS_IsAny<T> = boolean extends (T extends never ? true : false) ? true : false;
 type __VLS_PickNotAny<A, B> = __VLS_IsAny<A> extends true ? B : A;
@@ -19,7 +19,7 @@ type __VLS_Prettify<T> = {
 
 type __VLS_GlobalComponents =
 	// @ts-ignore
-	& __VLS_PickNotAny<import('vue').GlobalComponents & {}, {}>
+	__VLS_PickNotAny<import('vue').GlobalComponents & {}, {}>
 	// @ts-ignore
 	& __VLS_PickNotAny<import('@vue/runtime-core').GlobalComponents & {}, {}>
 	// @ts-ignore

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -18,16 +18,16 @@ type __VLS_Prettify<T> = {
 } & {};
 
 // @ts-ignore
-type VueGlobalComponents = import('vue').GlobalComponents;
+type __VLS_VueGlobalComponents = import('vue').GlobalComponents;
 // @ts-ignore
-type VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalComponents;
+type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalComponents;
 // @ts-ignore
-type VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
+type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
 type __VLS_GlobalComponents =
-	& __VLS_PickNotAny<VueGlobalComponents & VueGlobalComponents, {}>
-	& __VLS_PickNotAny<VueRuntimeCoreGlobalComponents & VueRuntimeCoreGlobalComponents, {}>
-	& __VLS_PickNotAny<VueRuntimeDomGlobalComponents & VueRuntimeDomGlobalComponents, {}>
+	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>
+	& __VLS_PickNotAny<__VLS_VueRuntimeCoreGlobalComponents & __VLS_VueRuntimeCoreGlobalComponents, {}>
+	& __VLS_PickNotAny<__VLS_VueRuntimeDomGlobalComponents & __VLS_VueRuntimeDomGlobalComponents, {}>
 	& Pick<typeof import('${vueCompilerOptions.lib}'),
 		// @ts-ignore
 		'Transition'

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -24,7 +24,7 @@ type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalCo
 // @ts-ignore
 type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
-// Hack: `Type & Type` is a hack to make unresolved imports' type work.
+// Hack: \`Type & Type\` is a hack to make unresolved imports' type work.
 // Ref: #3274
 type __VLS_GlobalComponents =
 	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -25,19 +25,19 @@ type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalCo
 // @ts-ignore
 type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
-// Hack: \`Type & Type\` is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
+// Hack: Type & Type is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
 // Since typescript 5.1.3, the behavior of unresolved imports has changed.
 // For example, the following code:
-// \`\`\`ts
+//
 // type A = import('foo').NotExist extends any ? true : false;
 // type B = A extends any ? true : false;
-// \`\`\`
-// In typescript 5.0.4, A and B are both \`true\`.
-// However, in 5.1.3, A becomes \`any\` and B becomes \`any\` too!
-// Actually, any type which includes unresolved imports will be turned into \`any\`.
+//
+// In typescript 5.0.4, A and B are both true.
+// However, in 5.1.3, A becomes any and B becomes any too!
+// Actually, any type which includes unresolved imports will be turned into any.
 // Even only a reference, like B.
 // So we need to turn it into a \`real any\` and avoid pollution.
-// This hack does it by intersect itself with itself 
+// This hack does it by intersect itself with itself.
 // Ref: #3274
 type __VLS_GlobalComponents =
 	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -17,32 +17,13 @@ type __VLS_Prettify<T> = {
 	[K in keyof T]: T[K];
 } & {};
 
-// TODO: revert this once microsoft/TypeScript#54630 is fixed
-// @ts-ignore
-type __VLS_VueGlobalComponents = import('vue').GlobalComponents;
-// @ts-ignore
-type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalComponents;
-// @ts-ignore
-type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
-
-// Hack: Type & Type is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
-// Since typescript 5.1.3, the behavior of unresolved imports has changed.
-// For example, the following code:
-//
-// type A = import('foo').NotExist extends any ? true : false;
-// type B = A extends any ? true : false;
-//
-// In typescript 5.0.4, A and B are both true.
-// However, in 5.1.3, A becomes any and B becomes any too!
-// Actually, any type which includes unresolved imports will be turned into any.
-// Even only a reference, like B.
-// So we need to turn it into a real any and avoid pollution.
-// This hack does it by intersect itself with itself.
-// Ref: #3274
 type __VLS_GlobalComponents =
-	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>
-	& __VLS_PickNotAny<__VLS_VueRuntimeCoreGlobalComponents & __VLS_VueRuntimeCoreGlobalComponents, {}>
-	& __VLS_PickNotAny<__VLS_VueRuntimeDomGlobalComponents & __VLS_VueRuntimeDomGlobalComponents, {}>
+	// @ts-ignore
+	& __VLS_PickNotAny<import('vue').GlobalComponents & {}, {}>
+	// @ts-ignore
+	& __VLS_PickNotAny<import('@vue/runtime-core').GlobalComponents & {}, {}>
+	// @ts-ignore
+	& __VLS_PickNotAny<import('@vue/runtime-dom').GlobalComponents & {}, {}>
 	& Pick<typeof import('${vueCompilerOptions.lib}'),
 		// @ts-ignore
 		'Transition'

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -36,7 +36,7 @@ type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComp
 // However, in 5.1.3, A becomes any and B becomes any too!
 // Actually, any type which includes unresolved imports will be turned into any.
 // Even only a reference, like B.
-// So we need to turn it into a \`real any\` and avoid pollution.
+// So we need to turn it into a real any and avoid pollution.
 // This hack does it by intersect itself with itself.
 // Ref: #3274
 type __VLS_GlobalComponents =

--- a/packages/vue-language-core/src/utils/directorySharedTypes.ts
+++ b/packages/vue-language-core/src/utils/directorySharedTypes.ts
@@ -24,7 +24,7 @@ type __VLS_VueRuntimeCoreGlobalComponents = import('@vue/runtime-core').GlobalCo
 // @ts-ignore
 type __VLS_VueRuntimeDomGlobalComponents = import('@vue/runtime-dom').GlobalComponents;
 
-// Hack: \`Type & Type\` is a hack to make unresolved imports' type work.
+// Hack: \`Type & Type\` is a hack to make unresolved imports' type to be "real any" in TS 5.1.3.
 // Ref: #3274
 type __VLS_GlobalComponents =
 	& __VLS_PickNotAny<__VLS_VueGlobalComponents & __VLS_VueGlobalComponents, {}>

